kernel/src/arch/x86_64/gdt.rs

```diff
@@ -1,29 +1,143 @@
-use x86_64_crate::instructions::segmentation::Segment;
-use x86_64_crate::instructions::segmentation::CS;
-use x86_64_crate::instructions::tables::load_tss;
-use x86_64_crate::structures::gdt::{Descriptor, GlobalDescriptorTable};
-use x86_64_crate::structures::tss::TaskStateSegment;
-use x86_64_crate::VirtAddr;
+use core::mem::{self, MaybeUninit};
 
 pub const DOUBLE_FAULT_IST_INDEX: u16 = 0;
-static mut TSS: TaskStateSegment = TaskStateSegment::new();
-static mut GDT: GlobalDescriptorTable = GlobalDescriptorTable::new();
+
+const KERNEL_RING: u8 = 0;
+const USERLAND_RING: u8 = 3;
+
+const KERNEL_CODE_SEG: Segment = Segment::new(1, KERNEL_RING);
+const KERNEL_DATA_SEG: Segment = Segment::new(2, KERNEL_RING);
+const USER_CODE_SEG: Segment = Segment::new(3, USERLAND_RING);
+const USER_DATA_SEG: Segment = Segment::new(4, USERLAND_RING);
+
+const MAX_ENTRIES: usize = 5;
 
 pub fn init() {
+    let gdt = Gdt::new()
+        .insert(KERNEL_CODE_SEG, GdtEntry::new(0, true, KERNEL_RING))
+        .insert(KERNEL_DATA_SEG, GdtEntry::new(0, false, KERNEL_RING))
+        .insert(USER_CODE_SEG, GdtEntry::new(0, true, USERLAND_RING))
+        .insert(USER_DATA_SEG, GdtEntry::new(0, false, USERLAND_RING));
+
+    // SAFETY: Gdt and Segments are valid
     unsafe {
-        TSS.interrupt_stack_table[DOUBLE_FAULT_IST_INDEX as usize] = {
-            const STACK_SIZE: usize = 4096;
-            static mut STACK: [u8; STACK_SIZE] = [0; STACK_SIZE];
+        gdt.load(KERNEL_CODE_SEG, KERNEL_DATA_SEG);
+    }
+}
+
+struct Segment(u64);
+
+impl Segment {
+    pub const fn new(offset: u64, privilege: u8) -> Segment {
+        Segment((offset << 3) | (privilege as u64 & 0b11))
+    }
+
+    pub const fn get_offset(&self) -> usize {
+        (self.0 >> 3) as usize
+    }
+}
+
+#[derive(Default, Debug)]
+#[repr(C)]
+struct GdtEntry {
+    seg_lim_low: u16,
+    base_addr_low: u16,
+    base_addr_mid: u8,
+    /// accessed : 1
+    /// writeable : 1
+    /// expand_down : 1
+    /// executable : 1
+    /// desc_type : 1
+    /// privilege : 2
+    /// seg_present : 1
+    access: u8,
+    /// seg_lim_high : 4
+    /// avl : 1
+    /// cs_size : 1
+    /// default_operation_size : 1
+    /// granularity : 1
+    flags: u8,
+    base_addr_high: u8,
+}
 
-            let stack_start = VirtAddr::from_ptr(&STACK);
-            stack_start + STACK_SIZE
+const WRITEABLE: u8 = 1 << 1;
+const EXECUTABLE: u8 = 1 << 3;
+const DESC_TYPE: u8 = 1 << 4;
+const PRIVILEGE_SHIFT: u8 = 5;
+const SEG_PRESENT: u8 = 1 << 7;
+const AVL: u8 = 1 << 4;
+const CS_SIZE: u8 = 1 << 5;
+const GRANULARITY: u8 = 1 << 7;
```

Comment:
I didn't know you could do that ! How can you directly call them from inside the impl ? I can't seem to use `use GdtEntry::*;` meaning I have to call GdtEntry::WRITEABLE everytime instead of WRITEABLE which is harder to read.